src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt

```diff
@@ -138,3 +179,17 @@ class RsExtractFunctionConfig private constructor(
     }
 }
 
+private fun Ty.types(): List<Ty> = if (typeParameterValues.values.isEmpty())
+    listOf(this)
+else
+    listOf(this) + typeParameterValues.values.flatMap { it.types() }
+
+private fun Ty.dependTypes(bounds: Map<Ty, List<Ty>>, last: List<Ty> = emptyList()): List<Ty> {
+    val types = bounds[this]
+    return if (types == null || types.isEmpty())
+        emptyList()
+    else
+        types + types.flatMap {
+            if (last.contains(it)) emptyList() else it.dependTypes(bounds, last + types)
```

Comment:
Let's avoid list concatenation in recursive function - it creates a lot of temporary objects and copy both lists for every `+`
If you want/need to collect something in recursive function you can use the following pattern:
```kotlin
private fun collectSomething(arg): SomeCollectionInterface {
    val collection = SomeMutableCollection()

    fun collect(args) {
        // do something and put result into collection
        // call recursively if needed
    }

    collect(arg)
    return collection
}
```